gap/oper.gi

```diff
@@ -1833,3 +1833,60 @@ function(gr, edges)
   od;
   return true;
 end);
+
+InstallMethod(AsSemigroup,
+"for a function and a digraph", [IsFunction, IsDigraph],
+function(filt, digraph)
+  local red, top, im, max, n, gens, i;
+
+  if not filt = IsPartialPermSemigroup then
+      ErrorNoReturn("Digraphs: AsSemigroup usage,\n",
+                    "the first argument must be IsPartialPermSemigroup,");
+  elif not IsJoinSemilatticeDigraph(digraph) then
+    if IsMeetSemilatticeDigraph(digraph) then
+      return AsSemigroup(IsPartialPermSemigroup, DigraphReverse(digraph));
+    else
+      ErrorNoReturn("Digraphs: AsSemigroup usage,\n",
+                    "the second argument must be a join semilattice ",
+                    "digraph or a meet semilattice digraph,");
+    fi;
+  fi;
+
+  red := DigraphReflexiveTransitiveReduction(digraph);
+  top := DigraphTopologicalSort(digraph);
+  # im[i] will store the image of the idempotent partial perm corresponding to
+  # vertex i of the arugment <digraph>
+  im         := [];
+  im[top[1]] := [];
+  max        := 1;
+
+  n := DigraphNrVertices(digraph);
+  # For each vertex, the corresponding idempotent has an image
+  # containing all images of idempotents below it.
+  for i in [2 .. n] do
+    im[top[i]] := Union(List(OutNeighboursOfVertex(red, top[i]), j -> im[j]));
+    # When there is only one neighbour, we must add a point to the image to
+    # distinguish the two idempotent partial perms.
+    if Length(OutNeighboursOfVertex(red, top[i])) = 1 then
+      Add(im[top[i]], max);
+      max := max + 1;
+    fi;
+  od;
+
+  # Determine a small generating set
+  gens := Filtered([1 .. n], j -> Size(InNeighboursOfVertex(red, j)) < 2);
+  return Semigroup(List(gens, g -> PartialPerm(im[g], im[g])));
+end);
+
+InstallMethod(AsMonoid,
+"for a function and a digraph", [IsFunction, IsDigraph],
+function(filt, digraph)
+  if not filt = IsPartialPermMonoid then
+      ErrorNoReturn("Digraphs: AsMonoid usage,\n",
+                    "the first argument must be IsPartialPermMonoid,");
+  elif not IsLatticeDigraph(digraph) then
+      ErrorNoReturn("Digraphs: AsMonoid usage,\n",
+                    "the second argument must be a lattice digraph,");
+  fi;
+  return AsSemigroup(IsPartialPermSemigroup, digraph);
```

Comment:
Is this definitely guaranteed to return something in the category `IsMonoid`?

Reply:
The second one here won't work, since `IsMonoid` is a category and you can't change categories that way. To be safe you could go with the first one of those options, but if you look at the GAP code and convince yourself that it will always return something in the category `IsMonoid` given the generators that correspond to a lattice digraph, then I don't mind if you leave it how it is.

Reply:
So when I apply AsSemigroup to a lattice digraph it prints `<partial perm monoid ... >` and satisfies `IsMonoid`. However the function `GeneratorsOfMonoid` returns a list containing the identity of the monoid. I would need to remove that somehow to claim the generating set is of minimal degree in the documentation.

```
gap> digraph := Digraph([[1],[1,2],[1,2,3]]);
<digraph with 3 vertices, 6 edges>
gap> S := AsSemigroup(IsPartialPermSemigroup, digraph);
<partial perm monoid of rank 2 with 3 generators>
gap> IsMonoid(S);
true
gap> GeneratorsOfSemigroup(S);
[ <empty partial perm>, <identity partial perm on [ 1 ]>, <identity partial perm on [ 1, 2 ]> ]
gap> GeneratorsOfMonoid(S);
[ <empty partial perm>, <identity partial perm on [ 1 ]>, <identity partial perm on [ 1, 2 ]> ]
 ```

Reply:
Yes, something is in the category `IsMonoid` if and only if `IsMonoid` returns `true`. It's just a technical thing: it means that for any element of the monoid you can call `One(x)` (or write `x^0`) and you get the identity element (or 'one') of that monoid. So you can have things that are technically semigroups with identity elements that do not satisfy `IsMonoid` in GAP.

So yeah, maybe it's a good idea to get the generators returned by `AsSemigroup(...)`, remove the identity, and then call `Monoid` on the resulting thing!

To clear something up – minimal degree is about the number of points on which the partial perms act on; minimal generating set means the least number of generators is used :)

Reply:
Thanks for clearing that up for me. I will do as you suggest with the generators.

> To clear something up – minimal degree is about the number of points on which the partial perms act on; minimal generating set means the least number of generators is used :)

Oops, I did know that about degree vs generating set. I really meant to say size not degree in my last comment!